apm-collector/apm-collector-ui/collector-ui-jetty-provider/src/test/java/org/apache/skywalking/apm/collector/ui/service/TopologyBuilderTest.java

```diff
@@ -0,0 +1,156 @@
+import org.apache.skywalking.apm.collector.storage.dao.ui.IApplicationReferenceMetricUIDAO;
+import org.apache.skywalking.apm.collector.storage.table.register.Application;
+import org.apache.skywalking.apm.collector.storage.ui.alarm.Alarm;
+import org.apache.skywalking.apm.collector.storage.ui.alarm.AlarmItem;
+import org.apache.skywalking.apm.collector.storage.ui.common.Duration;
+import org.apache.skywalking.apm.collector.storage.ui.common.Step;
+import org.apache.skywalking.apm.collector.storage.ui.common.Topology;
+import org.apache.skywalking.apm.collector.ui.utils.DurationUtils;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.mockito.internal.util.reflection.Whitebox;
+
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.mockito.Matchers.*;
+import static org.mockito.Mockito.mock;
+import static org.powermock.api.mockito.PowerMockito.when;
+
+/**
+ * @author lican
+ */
+public class TopologyBuilderTest {
+
+    private ApplicationCacheService applicationCacheService;
+    private ServerService serverService;
+    private SecondBetweenService secondBetweenService;
+    private AlarmService alarmService;
+    private TopologyBuilder topologyBuilder;
+    private Duration duration;
+    private long startSecondTimeBucket;
+    private long endSecondTimeBucket;
+    private long startTimeBucket;
+    private long endTimeBucket;
+
+    @Before
+    public void setUp() throws Exception {
+        ModuleManager moduleManager = mock(ModuleManager.class);
+        when(moduleManager.find(anyString())).then(invocation -> new MockModule());
+        topologyBuilder = new TopologyBuilder(moduleManager);
+        applicationCacheService = mock(ApplicationCacheService.class);
+        alarmService = mock(AlarmService.class);
+        secondBetweenService = mock(SecondBetweenService.class);
+        Whitebox.setInternalState(topologyBuilder, "applicationCacheService", applicationCacheService);
+        Whitebox.setInternalState(topologyBuilder, "alarmService", alarmService);
+        Whitebox.setInternalState(topologyBuilder, "secondBetweenService", secondBetweenService);
+        duration = new Duration();
+        duration.setEnd("2018-02");
+        duration.setStart("2018-01");
+        duration.setStep(Step.MONTH);
+        startSecondTimeBucket = DurationUtils.INSTANCE.startTimeDurationToSecondTimeBucket(duration.getStep(), duration.getStart());
+        endSecondTimeBucket = DurationUtils.INSTANCE.endTimeDurationToSecondTimeBucket(duration.getStep(), duration.getEnd());
+        startTimeBucket = DurationUtils.INSTANCE.exchangeToTimeBucket(duration.getStart());
+        endTimeBucket = DurationUtils.INSTANCE.exchangeToTimeBucket(duration.getEnd());
+    }
+
+    @Test
+    public void build() throws ParseException {
```

Comment:
@wu-sheng Real scenario of this test is very hard to mock. I suggest use automated test with the real segment data to test this method. So, this test case is enough.

Reply:
@peng-yongsheng You can feel relax about these type test cases. I agree, on one can be easy to mock topology and trace related test cases. We can count on your automatic backend test to cover these.